### PR TITLE
Seek into the beat map region when quantize play

### PR DESCRIPTION
**Pull request by daschuer**
This fixes https://bugs.launchpad.net/mixxx/+bug/1874918

The common case is that beats are matched when the track is cued short before the first beat. 

The downside is that it may result int a long seek if the track is cued into the lead in region. 
But I think the user can expect this behavior, if he ask Mixxx to match a beat. Now actually a beat is matched. He has always the option to disable quantize and keep the cued position.

### CODE REVIEW

File: src/engine/controls/bpmcontrol.cpp [hunk 1]
```diff
@@ -731,53 +731,23 @@ double BpmControl::getBeatMatchPosition(
         double dThisPosition, bool respectLoops, bool playing) {
     // Without a beatgrid, we don't know the phase offset.
     if (!m_pBeats) {
+        qDebug() << "############1";
```
Comment (Holzhaus) [inline src/engine/controls/bpmcontrol.cpp:734]: leftover from debugging?

Comment (daschuer) [reply src/engine/controls/bpmcontrol.cpp]: oh yes, this slipped though.

Comment (uklotzde) [reply src/engine/controls/bpmcontrol.cpp]: In this case I would recommend a force push. Keeping this code in the repo doesn't make any sense at all, even if it builds.

File: src/engine/controls/bpmcontrol.cpp [hunk 2]
```diff
@@ -790,6 +757,41 @@ double BpmControl::getBeatMatchPosition(
...
+                &dThisNextBeat,
+                &dThisBeatLength,
+                nullptr);
+        // now we either have a useful next beat or there is non
+        if (dThisNextBeat == -1) {
+            // We can't match the next beat, give up.
+            return dThisPosition;
+        }
+    } else {
+        // Here we within the current position, try to look up beat lenght.
```
Comment (ywwg) [inline src/engine/controls/bpmcontrol.cpp:785]: // We are between the previous and next beats so we can do a standard lookup

File: src/engine/controls/bpmcontrol.cpp [hunk 3]
```diff
@@ -800,6 +802,10 @@ double BpmControl::getBeatMatchPosition(
 
     double dOtherPosition = pOtherEngineBuffer->getExactPlayPos();
 
+    double dOtherPrevBeat;
+    double dOtherNextBeat;
+    double dOtherBeatLength;
+    double dOtherBeatFraction;
```
Comment (ywwg) [inline src/engine/controls/bpmcontrol.cpp:808]: for safety, can we initialize these to some useful value?  0 or -1 or whatever

**Review by ywwg — CHANGES_REQUESTED** [review]
I'll test this out.  I bet this fixes my not-quantizing issue, because I bet the tracks are being loaded at sample 0 instead of the cue point.

File: src/engine/controls/bpmcontrol.cpp (hunk 3 above)
Comment (daschuer) [reply src/engine/controls/bpmcontrol.cpp]: Oh Yes, this reveals also a bug that we use -1 below. Now I have cached that case as well.

**Comment by daschuer** [comment]
Done

**Comment by daschuer** [comment]
I have just squashed the changes as requested by @uklotzde.

**Comment by ywwg** [comment]
this works for me, thanks!  Is it possible to add a test for this that fails without the change applied?

**Comment by Holzhaus** [comment]
Please add this to the appropriate milestone and project board for 2.3 or 2.4.

**Comment by daschuer** [comment]
Done, Test is in place.

**Comment by ywwg** [comment]
looks like it needs an update from master

File: src/test/enginesynctest.cpp [hunk 4]
```diff
@@ -1768,3 +1768,33 @@ TEST_F(EngineSyncTest, ChangeBeatGrid) {
...
+    BeatsPointer pBeats2 = BeatsPointer(new BeatMap(*m_pTrack2, 44100));
+    // Add two beats at 120 Bpm
+    pBeats2->addBeat(44100 / 2);
+    pBeats2->addBeat(44100);
+    m_pTrack2->setBeats(pBeats2);
+
+    ControlObject::set(ConfigKey(m_sGroup1, "quantize"), 1.0);
+    ControlObject::set(ConfigKey(m_sGroup2, "quantize"), 1.0);
+
+    // Make Channel1 master to weed out any channel ordering issues.
```
Comment (ywwg) [inline src/test/enginesynctest.cpp:1786]: remove this comment, it doesn't apply because we now always process the master enginebuffer first

**Comment by daschuer** [comment]
Done.

**Comment by daschuer** [comment]
@ywwg: Can this be merged now?

**Comment by ywwg** [comment]
yes sorry for the delay, thanks!

**Comment by ywwg** [comment]
ugh sorry my test update broke this again, my bad.

**Comment by daschuer** [comment]
Done again. 
I had to revert your merge PR, because it was against master. i need to get used to the new 2.3 branch.

**Comment by ywwg** [comment]
This looks good to me!  github still says it won't merge though.  If you can fix commits, feel free to merge this yourself.  I'm not able to keep up fast enough to beat the conflicts

**Comment by ywwg** [comment]
sorry for the continuing lateness on my part :(

**Comment by daschuer** [comment]
Thank you for review.

**Pull request merged**